src/NuGet.Core/NuGet.Packaging/Rules/ReferencesInNuspecMatchRefAssetsRule.cs

```diff
@@ -0,0 +1,185 @@
+using NuGet.Frameworks;
+using NuGet.Packaging.Core;
+using NuGet.RuntimeModel;
+
+namespace NuGet.Packaging.Rules
+{
+    class ReferencesInNuspecMatchRefAssetsRule : IPackageRule
+    {
+        private string _addToRefFormat = AnalysisResources.ReferencesInNuspecAndRefFilesDontMatchWarningAddToRefListItemFormat;
+        private string _addToNuspecFormat = AnalysisResources.ReferencesInNuspecAndRefFilesDontMatchWarningAddToNuspecListItemFormat;
+        private string _addToNuspecNoTfmFormat = AnalysisResources.ReferencesInNuspecAndRefFilesDontMatchWarningAddToNuspecNoTfmListItemFormat;
+        public string MessageFormat => AnalysisResources.ReferencesInNuspecAndRefFilesDontMatchWarning;
+
+        public IEnumerable<PackagingLogMessage> Validate(PackageArchiveReader builder)
+        {
+            var refFiles = builder.GetFiles().Where(t => t.StartsWith(PackagingConstants.Folders.Ref));
+            var nuspecReferences = GetReferencesFromNuspec(builder.GetNuspec());
+            var missingItems = Compare(nuspecReferences, refFiles);
+            return GenerateWarnings(missingItems);
+        }
+
+        internal IDictionary<string, IEnumerable<string>> GetReferencesFromNuspec(Stream nuspecStream)
+        {
+            var nuspecReferences = new Dictionary<string, IEnumerable<string>>();
+            var packageNuspec = new NuspecReader(nuspecStream);
+            var nuspec = packageNuspec.Xml;
+            if (nuspec != null)
+            {
+                XNamespace name = nuspec.Root.Name.Namespace;
+                var targetFrameworks = nuspec.Descendants(XName.Get("{" + name.NamespaceName + "}references")).Elements().Attributes("targetFramework");
+                nuspecReferences = targetFrameworks.ToDictionary(k => NuGetFramework.Parse(k.Value).GetShortFolderName(), k => k.Parent.Elements().Attributes("file").Select(f => f.Value));
+                var filesWithoutTFM = nuspec.Descendants(XName.Get("{" + name.NamespaceName + "}references")).Elements().Attributes("file").Select(f => f.Value);
+                nuspecReferences.Add("any", filesWithoutTFM);
+            }
+            return nuspecReferences;
+        }
+
+        internal IEnumerable<MissingReference> Compare(IDictionary<string, IEnumerable<string>> nuspecReferences, IEnumerable<string> refFiles)
+        {
+            var missingReferences = new List<MissingReference>();
+            if (nuspecReferences.Count() != 0)
+            {
+                if (refFiles.Count() != 0)
+                {
+                    var filesByTFM = refFiles.Where(t => t.Count(m => m == '/') > 1).GroupBy(t => NuGetFramework.ParseFolder(t.Split('/')[1]).GetShortFolderName(), t => Path.GetFileName(t));
+                    var missingSubfolderInFiles = nuspecReferences.Keys.Where(t => !GetAllKeys(filesByTFM).Contains(t) && !NuGetFramework.ParseFolder(t).GetShortFolderName().Equals("unsupported") && !NuGetFramework.ParseFolder(t).GetShortFolderName().Equals("any"));
```

Comment:
Remember that LINQ is basically just syntactic sugar for a `foreach` loop, and anything that happens in method predicates run once per item in the enumerable. Therefore `GetAllKeys(filesByTFM)` is run once per item in `nuspecReferences`, and the code twice calls`NuGetFramework.ParseFolder(t).GetShortFolderName()`, meaning the same value is parsed twice and `GetShortFolderName` is called twice, but the value will be the same both times.

The practical impact here is minimal since we wouldn't expect either `nuspecReferences` or `filesByTFM` to be very long. But it's still good to avoid obvious performance issues, even when the impact isn't large. The practice will help write better code when performance can make a big difference.

Also, this line is very long. On my laptop screen it's wider than the resolution of the screen, and that's before you take into account the solution explorer window taking up some space. Such long expressions should be split over multiple lines to make it easy for everyone to read.

Reply:
So the reason I did this is because the `nuspecReferences` weren't being validated. So the code on line 61 is doing that validation inline. It's a little messy, so if you have any ideas to clean it up, feel free.